File: backend/src/main/java/com/mapmory/backend/location/Location.java
package com.mapmory.backend.location;

import com.mapmory.backend.country.Country;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;

@Entity
@Table(name = "location")
public class Location {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "country_id", nullable = false)
    private Country country;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "parent_id")
    private Location parent;

    @Column(name = "region_code", nullable = false, length = 10)
    private String regionCode;

    @Column(nullable = false, length = 100)
    private String name;

    @Enumerated(EnumType.STRING)
    @Column(name = "location_type", nullable = false, length = 20)
    private LocationType locationType;

    protected Location() {
    }

    private Location(
            Country country,
            Location parent,
            String regionCode,
            String name,
            LocationType locationType
    ) {
        this.country = country;
        this.parent = parent;
        this.regionCode = regionCode;
        this.name = name;
        this.locationType = locationType;
    }

    public static Location of(
            Country country,
            Location parent,
            String regionCode,
            String name,
            LocationType locationType
    ) {
        return new Location(country, parent, regionCode, name, locationType);
    }

    public Long getId() {
        return id;
    }
}

File: client/shared/src/commonMain/kotlin/com/mapmory/shared/domain/model/Location.kt
package com.mapmory.shared.domain.model

// 위치를 저장하기 위한 도메인
data class Location(
    val id: Long, // 지역 자체의 내부 ID
    val countryId: Long, // 이 지역이 속한 국가의 ID
    val parentId: Long?, // DISTRICT가 속한 PROVINCE
    // 국가: ISO 3166-1 alpha-2, 대한민국 시·도: ISO 3166-2, 시·군·구: 행정표준코드
    val regionCode: String,
    val name: String, // 지역명(ex) 서울특별시, 강남구)
    val type: LocationType, // 지역 단계
)

enum class LocationType {
    PROVINCE, // 시, 도 단위(서울특별시, 경기도)
    DISTRICT, // 시, 군, 구 단위(강남구, 수원시)
}
